stylo: Bug 1392161 - Introduce CSSPixelLength as computed::Length

BorisChiou opened the pull request:
These are the inter-dependent patches of bug 1392161. We want to handle
extreme small lengths carefully for some properties, such as transform, so we
shouldn't use |Au| as the computed value of specified::Length. Now, we introduce
a new type, CSSPixelLength, which is a wrapper of CSSFloat, and it is the
computed value of specified::Length, so we can keep the fractional part
of computed::Length.

---
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix [Bug 1392161](https://bugzilla.mozilla.org/show_bug.cgi?id=1392161).
- [X] These changes do not require tests because there is a wpt test for this, and I also add some others in Gecko.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/18381)
<!-- Reviewable:end -->

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/gecko/generated/bindings.rs, components/style/values/specified/transform.rs, components/style/values/specified/mod.rs, components/style/properties/gecko.mako.rs, components/style/values/computed/mod.rs and 6 more
 * @canaltinova: components/style/gecko/generated/bindings.rs, components/style/values/specified/transform.rs, components/style/values/specified/mod.rs, components/style/properties/gecko.mako.rs, components/style/values/computed/mod.rs and 6 more
 * @emilio: components/layout/fragment.rs, components/style/gecko/generated/bindings.rs, components/style/values/specified/transform.rs, components/style/values/specified/mod.rs, components/style/properties/gecko.mako.rs and 7 more

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!

BorisChiou commented:
r?@emilio

Assume handling extremely small lengths for transform is necessary, so we should avoid using ```Au``` as the computed lengths of transform. The current implementation has many duplicated codes, but I don't have a good idea to reduce them. @nox has a good idea?

also cc @upsuper

nox commented:
As long as it works it's fine I guess. I don't currently have the mental bandwidth to think about how to derive more of this.

BorisChiou commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 65041874dfcc9002c74e62e85fdf051f103d17fc with merge 33c92bdb66c922cfadb13de4e81206f97bda89bc...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/5903)

emilio commented on components/style/values/specified/length.rs:
```diff
@@ -116,22 +149,31 @@ impl FontRelativeLength {
-                    FontMetricsQueryResult::NotAvailable => reference_font_size.scale_by(0.5 * length),
-                }
+                    FontMetricsQueryResult::NotAvailable => {
+                        FontReferenceSize::Half(reference_font_size)
```
Why is this needed? Can't we return just `reference_font_size * 0.5` here?

BorisChiou replied:
Yes. Thanks. Actually we don't really need it. I just want to use a type to represent the reference size. I will drop it.

emilio commented on components/style/values/specified/length.rs:
```diff
@@ -144,12 +186,13 @@ impl FontRelativeLength {
+                            FontReferenceSize::Full(reference_font_size)
                         } else {
-                            reference_font_size.scale_by(0.5 * length)
+                            FontReferenceSize::Half(reference_font_size)
```
Ditto, I don't get the point of `FontReferenceSize`, care to explain?

emilio commented on components/style/values/specified/length.rs:
```diff
@@ -282,6 +331,31 @@ impl AbsoluteLength {
+        const PX_PER_MM: CSSFloat = PX_PER_IN / 25.4;
+        const PX_PER_Q: CSSFloat = PX_PER_MM / 4.;
+        const PX_PER_PT: CSSFloat = PX_PER_IN / 72.;
+        const PX_PER_PC: CSSFloat = PX_PER_PT * 12.;
```
Can we remove the `AU_PER_FOO` constants and at least reuse this? If not, why not?

emilio commented on components/style/values/specified/transform.rs:
```diff
@@ -8,6 +8,7 @@ use cssparser::Parser;
 use parser::{Parse, ParserContext};
 use selectors::parser::SelectorParseError;
 use style_traits::{ParseError, StyleParseError};
+use super::Either;
```
Let's not use `super::Either`, just `values::Either`.

emilio commented on components/style/values/specified/transform.rs:
```diff
@@ -254,3 +255,39 @@ impl ToComputedValue for TimingFunction {
+
+impl Parse for TransformLengthOrPercentage {
+    #[inline]
+    fn parse<'i, 't>(context: &ParserContext,
```
Let's indent this properly.

emilio commented on components/style/values/specified/transform.rs:
```diff
@@ -254,3 +255,39 @@ impl ToComputedValue for TimingFunction {
+impl Parse for TransformLength {
+    #[inline]
+    fn parse<'i, 't>(context: &ParserContext,
+                     input: &mut Parser<'i, 't>) -> Result<Self, ParseError<'i>> {
```
ditto, let's use block indentation here.

emilio commented on components/style/values/computed/transform.rs:
```diff
@@ -170,3 +164,171 @@ impl TransformList {
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+#[animate(fallback = "Self::animate_fallback")]
+#[derive(Animate, Clone, Copy, Debug, PartialEq, ToAnimatedZero, ToCss)]
+pub enum TransformLengthOrPercentage {
```
This is still a lot of duplicated code :(

emilio reviewed (commented):
After seeing the other changes in https://bugzilla.mozilla.org/show_bug.cgi?id=1396692, I think the right way to do this is just using `CSSFloat` as the computed value everywhere (maybe with a new type, `CSSPixel`?), instead of duplicating all this code.

That'd solve the precision problem everywhere, is there any reason that can't be done?

nox commented:
I agree with @emilio. Maybe we don't need to fix the precision problem everywhere indeed, but if fixing it everywhere means we have less code to maintain, I'm all for it.

BorisChiou commented:
Thanks emilio. Thanks. it makes sense, and that's the idea way to do this. At first, I'm not sure the main purpose of using ```Au```. Thanks for clarification. I will land https://bugzilla.mozilla.org/show_bug.cgi?id=1396692 first (because it is much simpler), and then do replace Length with CSSPixelLength. :)

bors-servo commented:
:umbrella: The latest upstream changes (presumably #18409) made this pull request unmergeable. Please resolve the merge conflicts.

BorisChiou committed:
Replace Au with CSSPixelLength in CalcLengthOrPercentage.

We replace Au with CSSPixelLength for the length part of
computed::CalcLengthOrPercentage. Therefore, it would be easier to use
CSSPixelLength for all other LengthOrPercentage{*} types.

BorisChiou committed:
Fix extremely small pixel value for transform on Servo.

Sometimes, we want to use extremely small pixel value in transform, e.g.
translate(calc(0.001px)), which should be rendered correctly together
with scale(100000). This patch only fixes this case for Servo because
Stylo still uses Au on the Gecko side, even if we pass pixel value into
FFI directly in the previous patch.

BorisChiou committed:
Introduce CSSPixelLength and update NonNegativeLength.

First, we define computed::CSSPixelLength which contains a CSSFloat, a
pixel value, and then we replace computed::Length with CSSPixelLength.
Therefore, the |ComputedValue| of NoCalcLength, AbsoluteLength,
FontRelativeLength, ViewportPercentageLength, CharacterWidth, and
PhysicalLength is CSSPixelLength.

Besides, we drop NonNegativeAu, and replace computed::NonNegativeLength
with NonNegative<computed::Length>. (i.e. NonNegative<CSSPixelLength>)

BorisChiou committed:
Use CSSPixelLength in LengthOrPercentage{*}.

Replace Au with CSSPixelLength in LengthOrPercentage,
LengthOrPercentageOrAuto, and LengthOrPercentageOrNone.

BorisChiou commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 2b0bec50abd85fce29d613d94150112082ea2b2e with merge bab53c05682b37141f8dd253231daa9ffa15ea01...

BorisChiou commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit b9ee1dae724ac21a3c25eedb0d89640a40cc6854 with merge 78696a1287df7043e4887b029b2bf86a0da191d8...

BorisChiou commented:
gecko try: https://treeherder.mozilla.org/#/jobs?repo=try&revision=57ce027c52b4

BorisChiou commented on components/style/values/specified/length.rs:
```diff
@@ -209,10 +234,17 @@ impl ViewportPercentageLength {
-        // that add up to 100 does not overflow due to rounding differences
-        let trunc_scaled = ((length.0 as f64) * factor as f64 / 100.).trunc();
-        Au::from_f64_au(trunc_scaled)
+        // FIXME: Find a better way to drop the if-condition.
```
I added this during rebasing my patch. Manish fixed this for some rounding issues of viewport, but I need more time to check how to combine his patch with mine, so use this if-condition as a work-around for now.

BorisChiou commented on components/style/values/specified/length.rs:
```diff
@@ -209,10 +234,17 @@ impl ViewportPercentageLength {
+        } else {
+            // For the app unit value smaller then 1.0, we should keep its fraction part,
+            // So we don't truncate it.
+            CSSPixelLength::new((au / AU_PER_PX as f64) as f32)
```
I added this during rebasing my patch. Manish fixed this for some rounding issues of viewport, but I need more time to check how to combine his patch with mine, so use this if-condition as a work-around for now.

BorisChiou commented on components/style/values/specified/length.rs:
```diff
@@ -209,10 +234,17 @@ impl ViewportPercentageLength {
+            // See bug 989802. We truncate so that adding multiple viewport units
+            // that add up to 100 does not overflow due to rounding differences
+            CSSPixelLength::from(Au::from_f64_au(au.trunc()))
+        } else {
```
I added this during rebasing my patch. Manish fixed this for some rounding issues of viewport, but I need more time to check how to combine his patch with mine, so use this if-condition as a work-around for now.

BorisChiou commented on components/style/values/specified/length.rs:
```diff
@@ -209,10 +234,17 @@ impl ViewportPercentageLength {
+        if au > 1. {
+            // See bug 989802. We truncate so that adding multiple viewport units
+            // that add up to 100 does not overflow due to rounding differences
+            CSSPixelLength::from(Au::from_f64_au(au.trunc()))
```
I added this during rebasing my patch. Manish fixed this for some rounding issues of viewport, but I need more time to check how to combine his patch with mine, so use this if-condition as a work-around for now.

BorisChiou replied:
OK, it seems this break other test, so I should update this now. :(

BorisChiou commented:
If gecko try looks good, I will land this ASAP.

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/5975)

BorisChiou commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 08119d3adaa44b1483e67fb6d762356339b405e6 with merge 24c09e6122e1c790751b4d91b84302f1cac10b82...

BorisChiou commented:
Gecko try: https://treeherder.mozilla.org/#/jobs?repo=try&revision=e930cba9ad80

BorisChiou commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit ce9f1ed1438dd07422fe93962bdc0bdecd387c9a with merge ac1b49b1987d37b3c1ba67213b43c6d5fc373579...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/1708)

BorisChiou commented:
@bors-servo r=emilio p=1

bors-servo commented:
:pushpin: Commit ce9f1ed has been approved by `emilio`

<!-- @bors-servo r=emilio ce9f1ed1438dd07422fe93962bdc0bdecd387c9a -->

bors-servo commented:
:zap: Previous build results for [android](http://build.servo.org/builders/android/builds/8860), [arm32](http://build.servo.org/builders/arm32/builds/9094), [arm64](http://build.servo.org/builders/arm64/builds/9125), [linux-dev](http://build.servo.org/builders/linux-dev/builds/9184), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/5978), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/6031), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/9227), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/1682), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/5922), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/5980), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/1704), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/4481) are reusable. Rebuilding only [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/1688), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/1708)...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/8860), [arm32](http://build.servo.org/builders/arm32/builds/9094), [arm64](http://build.servo.org/builders/arm64/builds/9125), [linux-dev](http://build.servo.org/builders/linux-dev/builds/9184), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/5978), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/6031), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/9227), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/1682), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/1688), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/5922), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/5980), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/1709), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/1704), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/4481)
Approved by: emilio
Pushing ac1b49b1987d37b3c1ba67213b43c6d5fc373579 to master...
